core/server/master/src/main/java/alluxio/master/DefaultSafeModeManager.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master;
+
+import alluxio.Configuration;
+import alluxio.Constants;
+import alluxio.PropertyKey;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Manages safe mode state for Alluxio master.
+ */
+public class DefaultSafeModeManager implements SafeModeManager {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultSafeModeManager.class);
+
+  /**
+   * Safe mode state. The value will be null if master is not in safe mode, or a nanosecond time
```

Comment:
could we track this in milliseconds instead? We usually measure in milliseconds, and nanosecond resolution isn't required in this context

Reply:
I used nanosecond just so that I don't have to divided it every time. I am fine moving to millisecond if that is what we use everywhere else.

Reply:
`System.nanoTime()` is recommended for measuring elapsed time. `System.currentTimeMillis()` can be affected by any user or system initiated(leap second, time sync, etc.) change of wall clock time.

Reply:
That's really interesting, thanks for sharing! We've historically mostly used `currentTimeMillis()` - could you create a JIRA ticket for using `System.nanoTime()` instead?

Could you update the `#getCurrentMs` method in `CommonUtils` to use `System.nanoTime`, then use that? We should use `TimeUnit` to convert between units of time.

Reply:
Okay created ALLUXIO-3081 for tracking this problem.

I would not update `CommonUtils .getCurrentMs()` to use `System.nanoTime()` for now given that the value returned by`System.nanoTime()` is not based on the same start time as `System.currentTimeMillis()`. Per https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#nanoTime() we should only use `System.nanoTime()` for elapsed time calculation and should not mix its calculation with `currentTimeMillis()` or other wall clock time.  We need to review each use case of `CommonUtils .getCurrentMs()`and `System.currentTimeMillis()` independently.

Reply:
ok, sounds good. For now we can just change this class to use `TimeUnit` with `System.nanoTime()`